valhalla/mjolnir/elevationbuilder.h

```diff
@@ -16,6 +15,7 @@ class ElevationBuilder {
    * Add elevation information to the graph tiles.
    */
   static void Build(const boost::property_tree::ptree& pt);
+  static bool add_elevations(const std::string& tile, const boost::property_tree::ptree& pt);
```

Comment:
we should keep this to one method `Build` so that it matches with the rest of the builders. What logic do we have in `valhalla_add_elevation` that is missing from the existing `Build` function? whether we called it from a standalone run of `valhalla_add_elevation` or from within `valhalla_build_tiles` it should do the same thing right? if the problem is about unit testing, i would suggest a private header which we can include for unit tests. let me know if this makes sense

Reply:
well... they are a bit different as Build tries to load elevation tiles for all valhalla tiles from the directory which is specified in the config file but valhalla_add_elevation should load elevation tiles only for a single valhalla tile passed as a parameter (although we still need config)

Reply:
i mean they are structured slightly differently but they do the same thing. the only real difference is once is inside of valhalla_build_tiles as a step in many steps, and the other is inside of valhalla_add_elevation as a single step.

both of them accomplish the same goal, pretty much the exact same way. they loop over all the tiles and add elevation to them. i definitely understand why we want to run build elevation into the tiles separate from the tile building process, but i dont understand why its useful to specifiy the list tiles separately. why isnt specifying the tileset (via config) enough?

its hard to understand why you would want to only add elevation to certain tiles in a tile set and not others or why you'd want to have to write code that asks whether or not the tile is in the tileset before fetching the elevation data. maybe you explain this use-case somewhere and i missed it?

Reply:
@kevinkreiser the last part of the https://github.com/valhalla/valhalla/pull/3391#issuecomment-965057330 might answer your question. My intention is to split valhalla tiles into batches and cache the batch result when it's ready
However(!), it was a visible issue when we fectched useless over-the-ocean elevation tiles(1.5TB in total). As I noticed in the comment we might don't need it when we stop fetching those elevation tiles as the stage duration is linear to  the elevation tileset size

And as I noted in that comment - there's no need to push this binary to the public repo. We can probably move it to the private fork only as it is kind of beta(I haven't seen it in work yet)

Reply:
sorry, maybe im not describing this very well. i think having a mode where we can fetch the data on demand and only pull the tiles that intersect the graphtiles is what ive always been advocating for. no problem there. all im saying is that it would be cool if the binary `valhalla_add_elevation` would be very thin wrapper around `ElevationBuilder::Build` basically take all the logic out of `valhalla_add_elevation` and put it in the `Build` method so that regardless of when we add the elevation, whether its in `valhalla_build_tiles` or in a separate call via `valhalla_add_elevation` its using the same code path. and the way that it works would be determined completely by config. so if someone didnt configure a tile url for elevation then it would only add elevation data that it found already int he directory, but if they did configure a tile directory and a url then it would fetch that data and store it in the tile directory while it was adding elevation. i'm just trying to keep things simple and not add so many ways to do the same thing.

i hope this makes sense just let me know if not, and we can always just have a quick video chat. its always possible that im misunderstanding why we cant do it like this :smile:

Reply:
Sorry I haven't been following that closely, but will this on-demand fetching method work on systems with less storage? For example if I only have 200GB available for elevation and I want to process the whole planet? Does it require disk space for the entire set of elevation tiles accessed or will at cache them and remove as needed so new ones can be retrieved?

Reply:
for now automatic elevation disposal is not implemented but as for me this idea sounds really cool. maybe we could have optional parameter in our config for this

Reply:
thats how the tile caching works for graph tiles as well (LRU). yet another opportunity to share code